File: archive/pulumi-ts/Pr6224/lib/network-stack.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable quotes */
/* eslint-disable @typescript-eslint/quotes */
/* eslint-disable prettier/prettier */

/**
 * network-stack.ts
 *
 * Creates VPC with public and private subnets, NAT gateways, and internet gateway.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface NetworkStackArgs {
  environmentSuffix: string;
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class NetworkStack extends pulumi.ComponentResource {
  public readonly vpcId: pulumi.Output<string>;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];

  constructor(
    name: string,
    args: NetworkStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:network:NetworkStack', name, args, opts);

    const { environmentSuffix, tags } = args;

    // Create VPC
    const vpc = new aws.ec2.Vpc(
      `vpc-${environmentSuffix}`,
      {
        cidrBlock: '10.0.0.0/16',
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: {
          Name: `vpc-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create Internet Gateway
    const igw = new aws.ec2.InternetGateway(
      `igw-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `igw-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Get availability zones
    const azs = aws.getAvailabilityZonesOutput({
      state: 'available',
    });

    // Create 2 public subnets in different AZs
    const publicSubnet1 = new aws.ec2.Subnet(
      `public-subnet-1-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        cidrBlock: '10.0.1.0/24',
        availabilityZone: azs.names[0],
        mapPublicIpOnLaunch: true,
        tags: {
          Name: `public-subnet-1-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const publicSubnet2 = new aws.ec2.Subnet(
      `public-subnet-2-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        cidrBlock: '10.0.2.0/24',
        availabilityZone: azs.names[1],
        mapPublicIpOnLaunch: true,
        tags: {
          Name: `public-subnet-2-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create 2 private subnets in different AZs
    const privateSubnet1 = new aws.ec2.Subnet(
      `private-subnet-1-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        cidrBlock: '10.0.11.0/24',
        availabilityZone: azs.names[0],
        tags: {
          Name: `private-subnet-1-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const privateSubnet2 = new aws.ec2.Subnet(
      `private-subnet-2-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        cidrBlock: '10.0.12.0/24',
        availabilityZone: azs.names[1],
        tags: {
          Name: `private-subnet-2-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create Elastic IPs for NAT Gateways
    const eip1 = new aws.ec2.Eip(
      `nat-eip-1-${environmentSuffix}`,
      {
        domain: 'vpc',
        tags: {
          Name: `nat-eip-1-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const eip2 = new aws.ec2.Eip(
      `nat-eip-2-${environmentSuffix}`,
      {
        domain: 'vpc',
        tags: {
          Name: `nat-eip-2-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create NAT Gateways in public subnets
    const natGw1 = new aws.ec2.NatGateway(
      `nat-gw-1-${environmentSuffix}`,
      {
        allocationId: eip1.id,
        subnetId: publicSubnet1.id,
        tags: {
          Name: `nat-gw-1-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const natGw2 = new aws.ec2.NatGateway(
      `nat-gw-2-${environmentSuffix}`,
      {
        allocationId: eip2.id,
        subnetId: publicSubnet2.id,
        tags: {
          Name: `nat-gw-2-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Create public route table
    const publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `public-rt-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Add route to Internet Gateway
    new aws.ec2.Route(
      `public-route-${environmentSuffix}`,
      {
        routeTableId: publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: igw.id,
      },
      { parent: this }
    );

    // Associate public subnets with public route table
    new aws.ec2.RouteTableAssociation(
      `public-rta-1-${environmentSuffix}`,
      {
        subnetId: publicSubnet1.id,
        routeTableId: publicRouteTable.id,
      },
      { parent: this }
    );

    new aws.ec2.RouteTableAssociation(
      `public-rta-2-${environmentSuffix}`,
      {
        subnetId: publicSubnet2.id,
        routeTableId: publicRouteTable.id,
      },
      { parent: this }
    );

    // Create private route tables
    const privateRouteTable1 = new aws.ec2.RouteTable(
      `private-rt-1-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `private-rt-1-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const privateRouteTable2 = new aws.ec2.RouteTable(
      `private-rt-2-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: {
          Name: `private-rt-2-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Add routes to NAT Gateways
    new aws.ec2.Route(
      `private-route-1-${environmentSuffix}`,
      {
        routeTableId: privateRouteTable1.id,
        destinationCidrBlock: '0.0.0.0/0',
        natGatewayId: natGw1.id,
      },
      { parent: this }
    );

    new aws.ec2.Route(
      `private-route-2-${environmentSuffix}`,
      {
        routeTableId: privateRouteTable2.id,
        destinationCidrBlock: '0.0.0.0/0',
        natGatewayId: natGw2.id,
      },
      { parent: this }
    );

    // Associate private subnets with private route tables
    new aws.ec2.RouteTableAssociation(
      `private-rta-1-${environmentSuffix}`,
      {
        subnetId: privateSubnet1.id,
        routeTableId: privateRouteTable1.id,
      },
      { parent: this }
    );

    new aws.ec2.RouteTableAssociation(
      `private-rta-2-${environmentSuffix}`,
      {
        subnetId: privateSubnet2.id,
        routeTableId: privateRouteTable2.id,
      },
      { parent: this }
    );

    // Expose outputs
    this.vpcId = vpc.id;
    this.publicSubnetIds = [publicSubnet1.id, publicSubnet2.id];
    this.privateSubnetIds = [privateSubnet1.id, privateSubnet2.id];

    this.registerOutputs({
      vpcId: this.vpcId,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}

File: archive/cdktf-py/Pr7248/lib/network_stack.py
"""Network infrastructure for Blue-Green deployment"""
from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.nat_gateway import NatGateway
from cdktf_cdktf_provider_aws.eip import Eip
from cdktf_cdktf_provider_aws.route_table import RouteTable, RouteTableRoute
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation


class NetworkStack(Construct):
    """Network stack with VPC, subnets, and gateways"""

    # pylint: disable=redefined-builtin,too-many-instance-attributes
    def __init__(self, scope: Construct, id: str, environment_suffix: str):
        super().__init__(scope, id)

        # VPC
        self.vpc = Vpc(self, 'vpc',
            cidr_block='10.0.0.0/16',
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={'Name': f'bluegreen-vpc-v1-{environment_suffix}'}
        )

        # Public Subnets
        self.public_subnet_1 = Subnet(self, 'public_subnet_1',
            vpc_id=self.vpc.id,
            cidr_block='10.0.1.0/24',
            availability_zone='us-east-1a',
            map_public_ip_on_launch=True,
            tags={'Name': f'bluegreen-public-1-v1-{environment_suffix}'}
        )

        self.public_subnet_2 = Subnet(self, 'public_subnet_2',
            vpc_id=self.vpc.id,
            cidr_block='10.0.2.0/24',
            availability_zone='us-east-1b',
            map_public_ip_on_launch=True,
            tags={'Name': f'bluegreen-public-2-v1-{environment_suffix}'}
        )

        # Private Subnets
        self.private_subnet_1 = Subnet(self, 'private_subnet_1',
            vpc_id=self.vpc.id,
            cidr_block='10.0.10.0/24',
            availability_zone='us-east-1a',
            tags={'Name': f'bluegreen-private-1-v1-{environment_suffix}'}
        )

        self.private_subnet_2 = Subnet(self, 'private_subnet_2',
            vpc_id=self.vpc.id,
            cidr_block='10.0.11.0/24',
            availability_zone='us-east-1b',
            tags={'Name': f'bluegreen-private-2-v1-{environment_suffix}'}
        )

        # Internet Gateway
        self.igw = InternetGateway(self, 'igw',
            vpc_id=self.vpc.id,
            tags={'Name': f'bluegreen-igw-v1-{environment_suffix}'}
        )

        # EIP for NAT Gateway
        self.nat_eip = Eip(self, 'nat_eip',
            domain='vpc',
            tags={'Name': f'bluegreen-nat-eip-v1-{environment_suffix}'}
        )

        # NAT Gateway
        self.nat_gateway = NatGateway(self, 'nat_gateway',
            allocation_id=self.nat_eip.id,
            subnet_id=self.public_subnet_1.id,
            tags={'Name': f'bluegreen-nat-v1-{environment_suffix}'}
        )

        # Public Route Table
        self.public_rt = RouteTable(self, 'public_rt',
            vpc_id=self.vpc.id,
            route=[RouteTableRoute(
                cidr_block='0.0.0.0/0',
                gateway_id=self.igw.id
            )],
            tags={'Name': f'bluegreen-public-rt-v1-{environment_suffix}'}
        )

        # Associate public subnets with public route table
        RouteTableAssociation(self, 'public_rta_1',
            subnet_id=self.public_subnet_1.id,
            route_table_id=self.public_rt.id
        )

        RouteTableAssociation(self, 'public_rta_2',
            subnet_id=self.public_subnet_2.id,
            route_table_id=self.public_rt.id
        )

        # Private Route Table
        self.private_rt = RouteTable(self, 'private_rt',
            vpc_id=self.vpc.id,
            route=[RouteTableRoute(
                cidr_block='0.0.0.0/0',
                nat_gateway_id=self.nat_gateway.id
            )],
            tags={'Name': f'bluegreen-private-rt-v1-{environment_suffix}'}
        )

        # Associate private subnets with private route table
        RouteTableAssociation(self, 'private_rta_1',
            subnet_id=self.private_subnet_1.id,
            route_table_id=self.private_rt.id
        )

        RouteTableAssociation(self, 'private_rta_2',
            subnet_id=self.private_subnet_2.id,
            route_table_id=self.private_rt.id
        )

    @property
    def vpc_id(self):
        return self.vpc.id

    @property
    def public_subnet_ids(self):
        return [self.public_subnet_1.id, self.public_subnet_2.id]

    @property
    def private_subnet_ids(self):
        return [self.private_subnet_1.id, self.private_subnet_2.id]
